sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexAsyncClient.java

```diff
@@ -0,0 +1,732 @@
+public class SearchIndexAsyncClient {
+
+    /**
+     * The lazily-created serializer for search index client.
+     */
+    private static final SerializerAdapter SERIALIZER = initializeSerializerAdapter();
+
+    /**
+     * Search REST API Version
+     */
+    private final SearchServiceVersion apiVersion;
+
+    /**
+     * The endpoint for the Azure Cognitive Search service.
+     */
+    private final String endpoint;
+
+    /**
+     * The name of the Azure Cognitive Search index.
+     */
+    private final String indexName;
+
+    /**
+     * The logger to be used
+     */
+    private final ClientLogger logger = new ClientLogger(SearchIndexAsyncClient.class);
+
+    /**
+     * The underlying AutoRest client used to interact with the Azure Cognitive Search service
+     */
+    private final SearchIndexRestClientImpl restClient;
+
+    /**
+     * The pipeline that powers this client.
+     */
+    private final HttpPipeline httpPipeline;
+
+    /**
+     * Package private constructor to be used by {@link SearchIndexClientBuilder}
+     */
+    SearchIndexAsyncClient(String endpoint, String indexName, SearchServiceVersion apiVersion,
+        HttpPipeline httpPipeline) {
+
+        SearchServiceUrlParts parts = SearchServiceUrlParser.parseServiceUrlParts(endpoint);
+
+        if (StringUtils.isBlank(indexName)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid indexName"));
+        }
+        if (apiVersion == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid apiVersion"));
+        }
+        if (httpPipeline == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid httpPipeline"));
+        }
+
+        this.endpoint = endpoint;
+        this.indexName = indexName;
+        this.apiVersion = apiVersion;
+        this.httpPipeline = httpPipeline;
+
+        restClient = new SearchIndexRestClientBuilder()
+            .searchServiceName(parts.serviceName)
+            .indexName(indexName)
+            .searchDnsSuffix(parts.dnsSuffix)
+            .apiVersion(apiVersion.getVersion())
+            .pipeline(httpPipeline)
+            .serializer(SERIALIZER)
+            .build();
+    }
+
+    /**
+     * Gets the name of the Azure Cognitive Search index.
+     *
+     * @return the indexName value.
+     */
+    public String getIndexName() {
+        return this.indexName;
+    }
+
+    /**
+     * Gets the {@link HttpPipeline} powering this client.
+     *
+     * @return the pipeline.
+     */
+    HttpPipeline getHttpPipeline() {
+        return this.httpPipeline;
+    }
+
+    /**
+     * Uploads a collection of documents to the target index
+     *
+     * @param documents collection of documents to upload to the target Index.
+     * @return document index result.
+     */
+    public Mono<IndexDocumentsResult> uploadDocuments(Iterable<?> documents) {
```

Comment:
@srnagar in the case of this client, some of the APIs return PagedFluxBase, which causes the style check to fail on @ServiceMethod(returns = ReturnType.COLLECTION) anotation. How can this be mitigated?

Another issue is the annotation forces a change to some of the method names. Since the API already passed the API board review, I am reluctant to make naming changes as this stage.
@JonathanGiles @brjohnstmsft

Reply:
For ReturnType.COLLECTION, I believe the checkstyle checks for return type to be a `PagedFlux`. This client returns a `PagedFluxBase` instead which should be okay as long this was reviewed by the board (cc @JonathanGiles). So, I would suggest that we update checkstyle rule to allow both `PagedFlux` and `PagedFluxBase` to be returned for `ReturnType.COLLECTION`. 

Please file a GitHub issue to update checkstyle rules.

Reply:
Issue https://github.com/Azure/azure-sdk-for-java/issues/7499 filed. 
Currently parking this until issue is resolved, since the API passed API Board Review with current api method naming.